In [1]:
import psycopg2
import pandas as pd

C:\Users\Я\AppData\Local\Temp\ipykernel_18508\979458548.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [3]:
connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

Блок 3

In [4]:
query_3_1 = f'''select
                count(*)
        from public.VACANCIES
        '''

df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\1757471012.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

In [5]:
query_3_2 = f'''select
                count(*)
        from public.EMPLOYERS
        '''

df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\2300384298.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

In [6]:
query_3_3 = f'''select
                count(*)
        from public.AREAS
        '''

df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\3758333869.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

In [7]:
query_3_4 = f'''select
                count(*)
        from public.INDUSTRIES
        '''

df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\3689885850.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

В предоставленной базе данных 49197 вакансий от 23501 работодателей из 294 сфер деятельности, распределенные по 1362 регионам.

Блок 4

In [8]:
query_4_1 = f'''select 
                count(*),
                a.name
        from public.VACANCIES v
        left join public.AREAS a on v.area_id = a.id
        group by a.id
        order by 1 desc
        limit 5
        '''

df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\3305051054.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


count             name
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы

In [7]:
query_4_2 = f'''select
                count(*)
        from public.VACANCIES v
        where salary_from is not NULL or salary_to is not NULL
        '''

df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_16512\342209983.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

In [10]:
query_4_3 = f'''select
                'Нижняя граница' as "ЗП",
                round(avg(salary_from)) as "Среднее"
        from public.VACANCIES v
        union all
        select
                'Верхняя граница',
        round(avg(salary_to))
        from public.VACANCIES v
        '''

df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\3356419445.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


ЗП   Среднее
0   Нижняя граница   71065.0
1  Верхняя граница  110537.0

In [9]:
query_4_4 = f'''select
            count(*),
            schedule,
            employment
        from public.VACANCIES v
        group by schedule, employment
        order by 1 desc
        limit 1 offset 1
        '''

df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_16512\860927862.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


count          schedule        employment
0   7802  Удаленная работа  Полная занятость

In [12]:
query_4_5 = f'''select
            count(*),
            experience
        from public.VACANCIES v
        group by experience
        order by 1
        '''

df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\901307103.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

Ожидаемо большая часть вакансий сконцентрирована столицах крупнейших городах.
Лишь 20% работодателей указывают верхнюю и нижнюю границу ЗП, а половина не указывает их вообще. Из этого можно сделать вывод, что в этом вопросе работодателеи полагаются больше на собеседование.
Указанные границы в среднем от 70тр до 110тр. Так же видно, что работодателей больше интересуют сотрудники с полной занятостью, из которых подавляющее большинство на полный день. От 1 года до 3 лет - самая распространенная категория опыта работы, около половины вакансий

Блок 5

In [4]:
query_5_1 = f'''(select
            count(*),
            e.name
        from public.VACANCIES v
        left join public.EMPLOYERS e on v.employer_id = e.id
        group by e.id
        order by 1 desc
        limit 1)
        
        union all
        
        (select
            count(*),
            e.name
        from public.VACANCIES v
        left join public.EMPLOYERS e on v.employer_id = e.id
        group by e.id
        order by 1 desc
        limit 4 offset 1)
        '''

df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_14348\2623638309.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


count           name
0   1933         Яндекс
1    491     Ростелеком
2    444       Тинькофф
3    428           СБЕР
4    331  Газпром нефть

In [14]:
query_5_2 = f'''select
                count(e.id) as "Работодатели",
                count(v.id) as "Вакансии",
                a.name as "Регион"
        from public.AREAS a
        left join public.EMPLOYERS e on a.id = e.area
        left join public.VACANCIES v on a.id = v.area_id
        group by a.name
        having count(v.id) = 0
        order by count(e.id) desc
        limit 1
        '''

df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\1839552369.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Работодатели  Вакансии  Регион
0           410         0  Россия

In [15]:
query_5_3 = f'''select
                count(DISTINCT v.area_id) as "Число регионов",
                e.name
        from public.EMPLOYERS e
        join public.VACANCIES v on e.id = v.employer_id
        group by e.name
        order by 1 desc
        limit 1
        '''

df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\2721425585.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Число регионов    name
0             181  Яндекс

In [16]:
query_5_4 = f'''select
                count(*)
        from public.EMPLOYERS e
        left join public.EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
        where ei.industry_id is NULL
        '''

df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\2909173993.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

In [8]:
query_5_5 = f'''select
            count(ei.industry_id),
            e.name
        from public.EMPLOYERS e
        left join public.EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
        group by e.id
        having count(ei.industry_id) = 4
        order by 2
        limit 1 offset 2
    '''

df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_18508\3971531384.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


count  name
0      4  2ГИС

In [18]:
query_5_6 = f'''select
            count(employer_id)
        from public.EMPLOYERS_INDUSTRIES ei
        left join public.INDUSTRIES i on ei.industry_id = i.id
        where i.name = 'Разработка программного обеспечения'
    '''

df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_17084\4061446712.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

In [39]:
Em_id = 1740
# В задаче явно требовалось запарсить данные с википедии, но у меня не получилось, поэтому я их переписал
# ID Яндекса вынес отдельно, чтоб не присоединять таблицу с работодателями (для ускорения запроса)
query_5_7 = f'''
          with ds as(
               select a.name
               from public.VACANCIES v
               left join public.areas a on v.area_id = a.id
               where employer_id = {Em_id} and
                    (a.name = 'Москва' or
                    a.name = 'Санкт-Петербург' or
                    a.name = 'Новосибирск' or
                    a.name = 'Екатеринбург' or
                    a.name = 'Казань' or
                    a.name = 'Красноярск' or
                    a.name = 'Нижний Новгород' or
                    a.name = 'Челябинск' or
                    a.name = 'Уфа' or
                    a.name = 'Самара' or
                    a.name = 'Ростов-на-Дону' or
                    a.name = 'Краснодар' or
                    a.name = 'Омск' or
                    a.name = 'Воронеж' or
                    a.name = 'Пермь' or
                    a.name = 'Волгоград')
               )
          (select
               count(*),
               name
          from ds
          group by name)
          
          union all
          select
               count(*),
               'Total'
          from ds
     '''

df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_16512\1450011120.py:39: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


count             name
0      54           Москва
1      42  Санкт-Петербург
2      39     Екатеринбург
3      36  Нижний Новгород
4      35      Новосибирск
5      32          Воронеж
6      30        Краснодар
7      26              Уфа
8      26           Самара
9      25   Ростов-на-Дону
10     25            Пермь
11     25           Казань
12     24        Волгоград
13     23       Красноярск
14     22        Челябинск
15     21             Омск
16    485            Total

Большие компании публикуют вакансии с большим разбросом по регионам, например Яндекс у которого есть вакансии во всех крупных городах России. И с этим связаны странности, например, среди регионов есть "Россия", в котором много вакансий но нет резюме.
Так же у компаний может быть большой разброс по сферам деятельности, кто то их не указывает, у кого то их 5.

Блок 6

In [11]:
query_6_1 = f'''select 
            count(*)
        from public.VACANCIES v
        WHERE name ILIKE '%data%' 
            or name ILIKE '%данн%' 
    '''

df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_16304\2273871310.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

In [33]:
query_6_2 = f'''SELECT
            count(*)
        FROM public.VACANCIES v
        WHERE (v.name ILIKE '%data scientist%' 
            or v.name ILIKE '%data science%' 
            or v.name ILIKE '%исследователь данных%' 
            or v.name LIKE '%ML%' and v.name NOT LIKE '%HTML%' 
            or v.name ILIKE '%machine learning%' 
            or v.name ILIKE '%машинн%обучен%') 
            and (v.name ILIKE '%junior%' or v.experience = 'Нет опыта' or v.employment = 'Cтажировка')
    '''

df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_16304\1733955330.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

In [15]:
query_6_3 = f'''SELECT
            count(v.name) as "postgres&sql"
        FROM public.VACANCIES v
        WHERE (v.name ILIKE '%data scientist%'
            or v.name ILIKE '%data science%'
            or v.name ILIKE '%исследователь данных%'
            or v.name ILIKE '%ML%' and v.name NOT LIKE '%HTML%' and v.name not like '%QML%'
            or v.name ILIKE '%machine learning%'
            or v.name ILIKE '%машинн%обучен%')
            and(v.key_skills ILIKE '%postgres%'
            or v.key_skills ILIKE '%sql%')
    '''

df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_18508\1829044848.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


postgres&sql
0           229

In [12]:
query_6_4 = f'''SELECT
            count(v.name) as python
        FROM public.VACANCIES v
        WHERE (v.name ILIKE '%data scientist%' 
            or v.name ILIKE '%data science%' 
            or v.name ILIKE '%исследователь данных%' 
            or v.name ILIKE '%ML%' and v.name NOT LIKE '%HTML%' and v.name not like '%QML%' 
            or v.name ILIKE '%machine learning%' 
            or v.name ILIKE '%машинн%обучен%') 
            and v.key_skills ILIKE '%python%'
    '''

df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_18508\2371855553.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


python
0     357

In [36]:
query_6_5 = f'''SELECT
                round(avg(length(key_skills) - length(replace(key_skills, CHR(9),'')) + 1), 2)
        FROM public.VACANCIES
        WHERE (name ILIKE '%data scientist%' 
                or name ILIKE '%data science%' 
                or name ILIKE '%исследователь данных%' 
                or name LIKE '%ML%' and name NOT LIKE '%HTML%'
                or name ILIKE '%machine learning%' 
                or name ILIKE '%машинн%обучен%') 
                and key_skills IS NOT NULL
    '''
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_16304\2969496250.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


round
0   6.41

In [11]:
query_6_6 = f'''SELECT
            experience,
            round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from)) / 2)) as salary
            -- в случае, когда оба поля ненулевые, выполняется среднеарифметическое. А если одна из строк пустая, то она копирует значение полной и при делении получается значение полной
        FROM public.VACANCIES
        WHERE (name ILIKE '%data scientist%' 
            or name ILIKE '%data science%' 
            or name ILIKE '%исследователь данных%' 
            or name LIKE '%ML%' and name NOT LIKE '%HTML%'
            or name ILIKE '%machine learning%' 
            or name ILIKE '%машинн%обучен%') 
            and key_skills IS NOT NULL
            and (salary_from IS NOT NULL or salary_to IS NOT NULL)
            and experience = 'От 3 до 6 лет'
        group by experience
        '''
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Я\AppData\Local\Temp\ipykernel_18508\3866886743.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience    salary
0  От 3 до 6 лет  243115.0

Среди вакансий по DS довольно мало относящихся к уровню junior, что вполне логично, ведь в области требуют в среднем чуть больше 6 ключевых навыка, что довольно много для начинающего. Эта области имеет высокий порог вхождения, но но и оплата для опытных специалистов этому ссответствует